### PR TITLE
ipld: Test `GetSharesByNamespace` to check ordering

### PR DESCRIPTION
**Pull request by tzdybal**
testing(service/share): message->shares->IPLD roundtrip test

Resolves https://github.com/celestiaorg/celestia-node/issues/621

### CODE REVIEW

**Comment by Wondertan** [comment]
@tzdybal, Why is the block called `invalid`? Is it invalid or valid?

**Comment by Wondertan** [comment]
@tzdybal, so I run the test locally and seems like the data we get is ok, the problem with the ordering again

**Comment by liamsi** [comment]
Is this sth we could integrate with the ephemeral cluster? So instead of hard coding a test vector only, we would have a test that asserts this round trip works with every released versions of the stack? Cc @Bidon15 @jbowen93

**Comment by Wondertan** [comment]
Overkill

**Comment by jbowen93** [comment]
> Member

The

**Comment by jbowen93** [comment]
> Is this sth we could integrate with the ephemeral cluster? So instead of hard coding a test vector only, we would have a test that asserts these kind of round trips works with every released versions of the stack? Cc @Bidon15 @jbowen93

The Ephemeral cluster stands up a celestia-app node (currently multiple, see https://github.com/celestiaorg/ephemeral-cluster/issues/15), a celestia-node bridge node and a celestia-node light node. If there is an RPC that exposes the state of things then this could be tested, but I don't believe this is supported right now.

**Comment by Wondertan** [comment]
@tzdybal, Kindly ask you to co-author @vgonkivs in the commit with the fix.

**Comment by tzdybal** [comment]
> @tzdybal, Kindly ask you to co-author @vgonkivs in the commit with the fix.
Done :)

**Comment by liamsi** [comment]
We can move the discussion around testing into a separate issue but IMO:
 - the ephemeral cluster already spins up node as well and with testing ethermint it could/should also test retrieving data per namespace (as that ethermint instance will have its own namespace already)
 - integration aka swamp tests are good but orthogonal to the above; and so is the mock app (which will simplify integration tests a lot)

Regarding the cluster: we should only clarify when this is run/which events trigger this (only tagging new releases of particular repos or on each push to master, or on a regular basis etc). I defer to @Bidon15's and @jbowen93's judgement here.

File: service/share/share_test.go
Comment (liamsi) [inline service/share/share_test.go:328]: As all this contains specifics/internal of the tendermint block + commits, can we add a comment for future maintainers how to handle this test-vector in case there was a breaking change in tendermint (e.g. what if we switched to BLS, or, remove the partSetHeader etc).

**Review by liamsi — COMMENTED** [review]
Left one comment and I'll update the title. Otherwise looks really good!

File: service/share/share_test.go
Comment (liamsi) [reply service/share/share_test.go:328]: Both swamp, as well as e2e tests in the ephemeral cluster would make it unnecessary to hardcode test-vectors that contain and spill over implementation details of structs defined in other repos.

Comment (Wondertan) [reply service/share/share_test.go:328]: I don't like the fact that we need to copy the JSON structure here for the test, as we should be ideally making a real block with an integration test, but we are blocked on the mocked app, so this way turns out to be the simplest one atm.

**Review by Wondertan — APPROVED** [review]
LGTM!

**Comment by vgonkivs** [comment]
May be we can keep data from `sampleBlock` in json file and read from it in test?

**Comment by renaynay** [comment]
Maybe keep it in a directory `test_files` or something like that.

**Comment by vgonkivs** [comment]
`test_files` or `resources`

**Comment by renaynay** [comment]
`testfiles` is fine i think @vgonkivs

**Comment by Wondertan** [comment]
Wow, that is what we need. There is also a new feature in Go for embedding files: https://pkg.go.dev/embed. Pls use that

**Comment by Wondertan** [comment]
Plan let's not make a centralized dir for all the rest files. Ipld can have it's own dir, which the test should embed and rely upon.

**Comment by vgonkivs** [comment]
> Wow, that is what we need. There is also a new feature in Go for embedding files: https://pkg.go.dev/embed. Pls use that

The we should move on 1.18, no?

**Comment by Wondertan** [comment]
@vgonkiv, well, I already did that on my machine ;) But yeah, we should do that for the repo ideally. There are tons of things in 1.18 that we can gain from

**Comment by tzdybal** [comment]
Embed was introduced in [go 1.16](https://tip.golang.org/doc/go1.16#library-embed).

**Review by vgonkivs — APPROVED** [review]

**Pull request merged**
